### DISCUSSION TITLE
ARM Template - czy funkcja if niepoprawnie zwraca int?

### DISCUSSION THREAD

**Discussion by TomaszOledzki** [AzureCommunityPL/group#323] [Pomoc] [answered]
Cześć,

chyba źle korzystam z funkcji "if".

W docs widzę, że "if" po rozwiązaniu warunku potrafi zwrócić: string, int, object, or array:
https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-logical#if

No dobra, w takim razie robię następującego if-a dla property "maxSizeBytes":
```
"apiVersion": "2021-02-01-preview",
"type": "Microsoft.Sql/servers/databases"
...
properties: {
...
"maxSizeBytes": "[if(equals(parameters('env'),'dev'), 268435456000, 107374182400)]",
...
}
```

Czyli mamy takie warunek dla property "maxSizeBytes":
jeśli parameters('env') == 'dev' -> to zwróć int: 268435456000 -> co powinno dać:
`"maxSizeBytes": 268435456000`

jeśli parameters('env') != 'dev' -> to zwróć int: 107374182400 -> co powinno dać:
`"maxSizeBytes": 107374182400`

A Deployment rzuca mi coś tagiego w twarz:
```
##[error]Deployment template language expression evaluation failed: 'The language expression 'if(equals(parameters('env'),'dev'), 268435456000, 107374182400)' is not valid: the value '268435456000' at position '36' cannot be converted to number.'.
```

ARM Template reference dla SQL Database twierdzi, że wartość dla property "maxSizeBytes" ma być int.
https://docs.microsoft.com/en-us/azure/templates/microsoft.sql/servers/databases?tabs=json

Any ideas?

**Comment by ptrstpp950** [comment, marked as answer]
Obstawiam, acz głowy sobie uciąć nie dam, że number może być parsowany do int'a.
Czytaj jest to za duża liczba :)

p.s. Zwróć string i ma szansę zadziałać

**Reply by TomaszOledzki** [reply]
> Obstawiam, acz głowy sobie uciąć nie dam, że number może być parsowany do int'a.
> Czytaj jest to za duża liczba :)
> 
> p.s. Zwróć string i ma szansę zadziałać

Też tak obstawiam. Zrobiłem jako stringi i zadziałało.
`"maxSizeBytes": "[if(equals(parameters('env'),'dev'),'268435456000','107374182400')]",`

Chciałem elegancko, tj. zapisać string i skonwertować do int, ale też nie zabanglało.
`"maxSizeBytes": "[if(equals(parameters('env'),'dev'),int('268435456000'),int('107374182400'))]",`

Dziwne jest, że podanie value jako int działa:
`"maxSizeBytes": 268435456000`

**Reply by ptrstpp950** [reply]
`"maxSizeBytes": 268435456000` - to nie jest `int`, poniewaz jest za duży. To `long`, ale ponieważ w JSON nie ma to znaczenia to wartość w ciapkach i bez jak widać daje radę :)

**Reply by miqm** [reply]
Z tego co się orientuję to wewnętrznie Azure Resource Manager wewnętrznie korzysta głównie z int, czyli z 32bit. Podobny case był w bicep: https://github.com/Azure/bicep/issues/486 i tam ostatecznie liczby >int32 są transpilowane z użyciem funkcji `json`, np. `json('12345678901234567')`
